kotlinpoet/src/main/java/com/squareup/kotlinpoet/KmSpecs.kt

```diff
@@ -0,0 +1,405 @@
+        }
+        if (isOperator) {
+          addModifiers(KModifier.OPERATOR)
+        }
+        if (isInfix) {
+          addModifiers(KModifier.INFIX)
+        }
+        if (isInline) {
+          addModifiers(KModifier.INLINE)
+        }
+        if (isTailRec) {
+          addModifiers(KModifier.TAILREC)
+        }
+        if (isExternal) {
+          addModifiers(KModifier.EXTERNAL)
+        }
+        if (isExpect) {
+          addModifiers(KModifier.EXPECT)
+        }
+        if (isSuspend) {
+          addModifiers(KModifier.SUSPEND)
+        }
+        val returnTypeName = this@toFunSpec.returnType.toTypeName(typeParamResolver)
+        if (returnTypeName != UNIT) {
+          returns(returnTypeName)
+          addStatement(TODO_BLOCK)
+        }
+        receiverParameterType?.toTypeName(typeParamResolver)?.let { receiver(it) }
+      }
+      .tag(this)
+      .build()
+}
+
+@KotlinPoetKm
+private fun ImmutableKmValueParameter.toParameterSpec(
+  typeParamResolver: ((index: Int) -> TypeName)
+): ParameterSpec {
+  val paramType = varargElementType ?: type ?: throw IllegalStateException("No argument type!")
+  return ParameterSpec.builder(name, paramType.toTypeName(typeParamResolver))
+      .apply {
+        if (varargElementType != null) {
+          addModifiers(KModifier.VARARG)
+        }
+        if (isCrossInline) {
+          addModifiers(KModifier.CROSSINLINE)
+        }
+        if (isNoInline) {
+          addModifiers(KModifier.NOINLINE)
+        }
+        if (declaresDefaultValue) {
+          defaultValue(TODO_BLOCK)
+        }
+      }
+      .tag(this)
+      .build()
+}
+
+@KotlinPoetKm
+private fun ImmutableKmProperty.toPropertySpec(
+  typeParamResolver: ((index: Int) -> TypeName),
+  isConstructorParam: Boolean
+) = PropertySpec.builder(name, returnType.toTypeName(typeParamResolver))
+    .apply {
+      addVisibility { addModifiers(it) }
+      addModifiers(*flags.modalities
+          .filterNot { it == KModifier.FINAL && !isOverride }
+          .toTypedArray())
+      if (isOverride) {
+        addModifiers(KModifier.OVERRIDE)
+      }
+      if (isConst) {
+        addModifiers(KModifier.CONST)
+      }
+      if (isVar) {
+        mutable(true)
+      } else if (isVal) {
+        mutable(false)
+      }
+      if (isDelegated) {
+        delegate("") // Placeholder
+      }
+      if (isExpect) {
+        addModifiers(KModifier.EXPECT)
+      }
+      if (isExternal) {
+        addModifiers(KModifier.EXTERNAL)
+      }
+      if (isLateinit) {
+        addModifiers(KModifier.LATEINIT)
+      }
+      if (isConstructorParam || (!isDelegated && !isLateinit)) {
+        // TODO if hasConstant + elements, we could look up the constant initializer
+        when {
+          isConstructorParam -> initializer(name)
+          type.isNullable -> initializer("null")
+          else -> initializer(TODO_BLOCK)
+        }
+      }
+      if (hasGetter) {
+        val visibility = setterFlags.visibility
```

Comment:
Should this block be using `getterFlags`? Also, use it for both getter and setter to avoid duplication?

Reply:
yep! Good catch, will fix